src/services/SearchResults.cpp

```diff
@@ -230,21 +230,23 @@ void SearchResults::ProcessBlocksNibbles(const SearchResults& srSource, unsigned
     std::vector<unsigned int> vMatches;
     std::vector<unsigned char> vMemory;
     const unsigned int nPadding = Padding(m_nSize);
+    std::array<unsigned char, sizeof(MemBlock)> pPrev{};
 
     for (auto& block : srSource.m_vBlocks)
     {
         if (block.GetSize() > vMemory.capacity())
-            vMemory.reserve(block.GetSize());
+            vMemory.resize(block.GetSize());
 
-        unsigned char* pMemory = vMemory.data();
-        const unsigned char* pPrev = block.GetBytes();
+        
+        for (auto i = 0U; i < sizeof(MemBlock); i++)
+            pPrev.at(i) = block.GetByte(i);
```

Comment:
This is wrong. `pPrev` is a 16-byte buffer because that's what `sizeof(MemBlock)` is, but `block.GetSize()` may return much more than 16 bytes.
```
class MemBlock
{
...
        union // 8 bytes
        {
            unsigned char m_vBytes[8]{};
            unsigned char* m_pBytes;
        };

        unsigned int nAddress; // 4 bytes
        unsigned int nSize;    // 4 bytes
};
```
You're effectively copying 16-bytes from `m_pBytes`, regardless of how many bytes are available, or if `m_pBytes` is even used.

`MemBlock` is designed like short string optimization. If the block is 8 bytes or less, the internal buffer is used, otherwise memory is allocated on the heap. `pPrev` is a pointer to access the entire buffer of the previous block for comparing against the current state. As I mentioned a previous review - if you want to avoid the direct pointer logic, you need to call a method to get the nth byte instead of `pPrev[i]`. It looks like you added a `GetByte` method, but that only works if the buffer is less than 8 bytes. if `nSize` is greater than 8 bytes, the heap memory needs to be references. See the implementation of `GetBytes()`.

This should be dirt simple for you to test. Simply open the Memory Inspector for any game and hit "New 4-bit test" (`ProcessBlocksNibbles` is only called in 4-bit test mode). This will allocate a single block large enough to hold the entire RAM (i.e. 65536 bytes on NES) - much larger than the 16-bytes you're providing.

Additionally, you didn't make changes to the lambda functions passed to `ProcessBlocks`. I would expect the same problem/solution to be necessary there. There are unit tests for `ProcessBlocks` and large memory, but it doesn't look like there's any for `ProcessBlocksNibbles` with large memory.

Reply:
I'll see if it'll work with a polymorphic container so we don't have to specify size.